### PR TITLE
Beyond React 16: Time Slicing and Suspense API

### PR DESCRIPTION
**Pull request by unicodeveloper**

### CODE REVIEW

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  mail: prosper.otemuyiwa@auth0.com
+  avatar: https://en.gravatar.com/avatar/1097492785caf9ffeebffeb624202d8f?s=200
+tags:
+- react
+- reactjs
+- javascript
+- reactfiber
+- fiber
+- frontend
+- authentication
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: Add `time-slicing`, `suspense`, `fetcher`.

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+- 2017-10-26-whats-new-in-react16
+---
+
+---
+
+**TL;DR:** ReactJS is a UI library that's a very powerful tool amongst frontend developers in building JavaScript applications. In this article, I'll introduce you to a few features coming to ReactJS.
+
+---
+
+ReactJS is a JavaScript library, built and maintained by Facebook. As of today, it powers so many popular web and mobile platforms such as Twitter, Airbnb, Lyft, Dropbox, Pinterest, Whatsapp and Instagram. The latest release of ReactJS which is _React 16_ ships with a lot of features such as `Error Boundaries`, `Custom DOM Attributes definition`, `Fragments as return types`, `Portals` and so many others.
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: Remove `so`.

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+- authentication
+related:
+- 2017-02-21-reactjs-authentication-tutorial
+- 2017-01-24-optimizing-react
+- 2017-10-26-whats-new-in-react16
+---
+
+---
+
+**TL;DR:** ReactJS is a UI library that's a very powerful tool amongst frontend developers in building JavaScript applications. In this article, I'll introduce you to a few features coming to ReactJS.
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: Link to official React.js page: https://reactjs.org/

File: _posts/2018-03-01-timeslice-suspense-react16.markdown (hunk 2 above)
Comment (sebadoom) [reply _posts/2018-03-01-timeslice-suspense-react16.markdown]: Link to official React.js page: https://reactjs.org/

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+Making it easier for developers to build great user experiences using ReactJS has always been the goal of the ReactJS team. Building for great user experiences involves splitting the attention of developers into two distinct categories:
+
+* **Computing Power**
+* **Network Speed**
+
+With these categories spelt out, you start asking the following questions:
+
+- Are the users on a slow network? If so, how's the user experience? Can we (developers) control the loading states?
+- Are the users on a low-end device (devices with low CPU power)? If so, is using the app still a memorable experience effectively?
+- Are the users on a fast network? If so, is the experience seamless? no janky UI.
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: no -> No

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 5]
```diff
@@ -0,0 +1,174 @@
+---
+layout: post
+title: "Beyond React 16: Time Slicing and Suspense API"
+description: "Learn what's coming to ReactJS. Get a sneak peek of the powerful features that will grace React soon."
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: It appears the proper name for React is simply `React` not `ReactJS` (if this is the case, you should fix all references in this post).

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+## Time Slicing
+
+In Dan's talk, he said: _"We’ve built a generic way to ensure that high-priority updates like user input don’t get blocked by rendering low-priority updates"_. What does this mean? The ReactJS team named this concept **Time Slicing**. Let me explain in simpler terms.
+
+{% include tweet_quote.html quote_text="We’ve built a generic way to ensure that high-priority updates like user input don’t get blocked by rendering low-priority updates." %}
+
+ReactJS is concerned about a device's CPU power. While rendering, ReactJS ensures that it doesn't block the thread thus causing the app to freeze. 
+
+**Time-slicing** allows ReactJS, which now runs on _React Fiber_, to split computations of updates on children components into chunks during idle callbacks and rendering work is spread out over multiple frames. Now, during the process of asynchronous rendering, it ensures that if a user's device is very fast, updates within the app feel synchronous and if a user's device is slow, the app feels responsive. No freezing, No janky UI experience!
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: Is there anything that needs to be done to enable this? If not, say it.

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+- Are the users on a fast network? If so, is the experience seamless? no janky UI.
+- Are the users on a high-end device (devices with high CPU power)? If so, is the rendering flawless?
+
+These are valid questions that need answers. Let's explore how **Time Slicing** and **Suspense** in ReactJS can help deliver the best user experience for everyone.
+
+## Time Slicing
+
+In Dan's talk, he said: _"We’ve built a generic way to ensure that high-priority updates like user input don’t get blocked by rendering low-priority updates"_. What does this mean? The ReactJS team named this concept **Time Slicing**. Let me explain in simpler terms.
+
+{% include tweet_quote.html quote_text="We’ve built a generic way to ensure that high-priority updates like user input don’t get blocked by rendering low-priority updates." %}
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: Put something in front of the quote, otherwise this may look like something we did. Like this: `Dan Abramov: "...."`

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+**Time-slicing** allows ReactJS, which now runs on _React Fiber_, to split computations of updates on children components into chunks during idle callbacks and rendering work is spread out over multiple frames. Now, during the process of asynchronous rendering, it ensures that if a user's device is very fast, updates within the app feel synchronous and if a user's device is slow, the app feels responsive. No freezing, No janky UI experience!
+
+## Suspense
+
+In Dan's talk, he said: "We have built a generic way for components to suspend rendering while they load asynchronous data". 
+
+The simple definition of the **suspense** feature is that ReactJS can pause any state update until the data been fetched is ready to be rendered. In essence, ReactJS suspends the component tree while waiting for the data to be fetched completely. During the suspension, it goes ahead to handle other high-priority updates. 
+
+{% include tweet_quote.html quote_text="We have built a generic way for components to suspend rendering while they load asynchronous data." %}
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: Put something in front of the quote, otherwise this may look like something we did. Like this: `Dan Abramov: "...."`

File: _posts/2018-03-01-timeslice-suspense-react16.markdown [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+```
+_GraphQL query_
+
+In the example above, there is an `asyncMode` prop attached to the `Query component` that allows for async rendering. The sample code below shows the render method in the `Query` component.
+
+```js
+....
+if (loading) {
+    if(this.props.asyncMode) {
+        throw this.state.queryObservable!.result();
```
Comment (sebadoom) [inline _posts/2018-03-01-timeslice-suspense-react16.markdown]: Perhaps a short description of what the odd use of `!` here means would be a good idea.

File: _posts/2018-03-01-timeslice-suspense-react16.markdown (hunk 5 above)
Comment (unicodeveloper) [reply _posts/2018-03-01-timeslice-suspense-react16.markdown]: For SEO, we usually use `ReactJS`. That's what gonzalez recommended for the previous React articles I wrote.
